docs/proposals/20201214-bootstrap-failure-detection.md

```diff
@@ -0,0 +1,149 @@
+
+#### VM runcmd Pros:
+- Relatively simple; we already have mature Azure SDK patterns in the AzureMachine controller implementation, would not be that much additional work to incorporate a runcmd operation against node vms
+
+#### VM runcmd Cons:
+- Limited stdout from runcmd request output
+- As with the VM Boot Diagnostics solution, this requires the AzureMachine controller to actually synthesize the runcmd (or multiple runcmd) result(s) into a terminal state outcome
+
+### Option 5: Custom capz-specific Azure VM Extension (recommended)
+A custom Azure VM Extension is basically a unit of foo that does a very finite set of things on a VM as it bootstraps itself. We could use this to implement a set of capz-focused bootstrap failure reporting, to support both investigation and remediation.
+
+#### Custom capz VM Extension Pros:
+- The same generic flexibility (although rendered as a concrete solution, not at runtime) as the Custom Script Extension option above, but maintains availability for future, user-configurable Custom Script Extensions solutions
+- Exposes a convenient binary success/failure as part of the Azure VM resource itself
+- Easy to query from the AzureMachine controller
+- Easy for users to introspect via Azure APIs, CLI, portal
+- Named property allows for convenient disambiguation from generic Azure (or other non-capz-related) errors
+
+#### Custom capz VM Extension Cons:
+- Non-trivial (though one-time) administrative overhead to create a named Azure VM Extension
+- Ongoing maintenance requires a separate release workflow compared to capz (in other words, we don’t simply ship changes to this *with* capz releases)
+
+### Option 6: postKubeadmCommand
+This is an option that could be used for kubeadm bootstrap-provided solutions only, which exposes an array interface to execute a set of arbitrary, serialized shell statements (essentially a thin wrapper around cloud-init’s runcmd interface) after kubeadm finishes.
+
+#### postKubeadmCommand Pros:
+- The interface is already present, assuming we only want to solve this for the kubeadm bootstrap provider
+
+#### postKubeadmCommand Cons:
+- *only* works for kubeadm bootstrap provider
+- Sort of breaks the UX contract for postKubeadmCommand, which is intended to be a user-configurable interface and not reserved for use by cluster-api controllers
+
+## Conclusions
+A few conclusions surfaced when exploring these options:
+
+1. Evaluating simple success/failure of VM bootstrapping is most easily done on the VM itself, because under no scenarios is there an option *not* to source some of the relevant input data from the VM. And because we can’t avoid basically establishing a connection to the VM’s filesystem, it simplifies things greatly to do that locally via a process/daemon running on the VM.
```

Comment:
```suggestion
1. Evaluating simple success/failure of VM bootstrapping is most easily done on the VM itself, because under no scenarios is there an option *not* to source some of the relevant input data from the VM. And because we can’t avoid establishing a connection to the VM’s filesystem, it simplifies things greatly to do that locally via a process/daemon running on the VM.
```